src/interpreter/plugin/ComplexPlugin.ts

```diff
@@ -0,0 +1,422 @@
+      ],
+    },
+    'IMCOT': {
+      method: 'imcot',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMCSC': {
+      method: 'imcsc',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMCSCH': {
+      method: 'imcsch',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMSEC': {
+      method: 'imsec',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMSECH': {
+      method: 'imsech',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMSIN': {
+      method: 'imsin',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMSINH': {
+      method: 'imsinh',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMTAN': {
+      method: 'imtan',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMDIV': {
+      method: 'imdiv',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMPRODUCT': {
+      method: 'improduct',
+      parameters: [
+        { argumentType: ArgumentTypes.ANY },
+      ],
+      repeatLastArgs: 1,
+    },
+    'IMSUM': {
+      method: 'imsum',
+      parameters: [
+        { argumentType: ArgumentTypes.ANY },
+      ],
+      repeatLastArgs: 1,
+    },
+    'IMSUB': {
+      method: 'imsub',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMEXP': {
+      method: 'imexp',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMLN': {
+      method: 'imln',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMLOG10': {
+      method: 'imlog10',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMLOG2': {
+      method: 'imlog2',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+    'IMPOWER': {
+      method: 'impower',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+        { argumentType: ArgumentTypes.NUMBER },
+      ],
+    },
+    'IMSQRT': {
+      method: 'imsqrt',
+      parameters: [
+        { argumentType: ArgumentTypes.COMPLEX },
+      ],
+    },
+  }
+
+  public complex(ast: ProcedureAst, formulaAddress: SimpleCellAddress): InternalScalarValue {
+    return this.runFunction(ast.args, formulaAddress, this.metadata('COMPLEX'),
+      (re: number, im: number, unit: string) => {
+        if(unit !== 'i' && unit !== 'j') {
+          return new CellError(ErrorType.VALUE, ErrorMessage.ShouldBeIorJ)
+        }
+        return coerceComplexToString([re, im], unit)
```

Comment:
I know its tempting, but it will be hard to do do without some ugly hacking in the engine. When we return the value from function, we need to know whether it is going to be used as a string or complex value, which depends on the 'above' function.